### ISSUE TITLE
cmd/gosh: $@ fails in Arithmetic Expansion

### ISSUE THREAD

**Issue by rob-myers** [mvdan/sh#317]
```sh
~$ gosh
$ foo () { echo $(( $@ )); }
$ foo 0
0 # correct
$ foo 1 + 1
0 # should be 2
$ foo 1
# program hangs
```

**Comment by rob-myers** [comment]
In `bash` it seems that `$(( x ))` evaluates as if assigning to a `declare -i` variable.
- `declare -i x; x='2 +'` gives the same error message as you mentioned.
- `declare -i x; x='1+1'` and `y='1+1'; x=$(( y ))` both assign `x=2`.

Perhaps this isn't weird behaviour?
It conforms to POSIX, and interprets 'undollared' variables in a natural way.

Looking at POSIX, it seems to require `$(( $@ ))` to act as I expected.
But it's not from a real program. I came across it while trying to write my own bash interpreter.

**Comment by mvdan** [comment]
Yes, POSIX does require this. But if we conformed to POSIX completely, there's a whole lot that the syntax package would never do. For example, `$((expression))` would be parsed as just that, like a single word. Only after expanding its body would we actually parse the arithmetic expression.

The smallest step we could take is allow `a='1+1'; echo $((a))` and `a='1+1'; echo $(($a))` to work. This requires zero `syntax` changes, and only requires `interp` to recursively parse arithmetic expressions when expanding variables.

The bigger step would be to allow `a='1+'; echo $(($a 2))`. This would require the `syntax` package to not parse arithmetic expressions upfront at all, as I described in the first paragraph.

The second step would bring us in line with POSIX. However, note that POSIX shell and Bash are unreasonable at times. For example, see https://github.com/mvdan/sh#caveats for some other edge cases that are not supported on purpose. Taking the second step would make the `syntax` package much less useful and a bit more complex.

For now I've taken the direction of parsing arithmetic expressions completely statically. I could take the first step outlined above; that would not be a considerable change, and it would likely solve most of the major edge cases like the one you suggested. I'm not keen on taking the second route, unless there is a very good reason to do so.

Compatibility with POSIX/Bash is certainly a major decision maker, but it's not the only one. I don't think this edge case is as clear as "no backtracking in the parser", but I still think we want to break with POSIX here.

**Comment by mvdan** [comment]
> I came across it while trying to write my own bash interpreter.

I'm also curious - are you implementing your own interpreter from scratch? There's one in the `expand` and `interp` packages already, I just haven't gotten to exposing it in `mvdan-sh` yet :)

**Comment by rob-myers** [comment]
Thanks for reiterating that POSIX interprets arithmetic at runtime -- I hadn't realised. I much prefer your static parsing approach.

I'd assumed (naively) that since `ArithmExpr` can be a `Word`, and `Word` has `Parts`, one could represent the case `$(( a 2 ))` as e.g. `ArithmExp( Word [Lit a,Lit \s,Lit 2] )`. At runtime one would have to evaluate `a`, concatenate, re-parse, then evaluate. I don't claim this is a good idea.

I'm aware of your interpreter (although my golang is not good). I'm writing my own to graphically animate bash in a very fine-grained way. E.g. a single step might animate the evaluation of a binary arithmetic operation (after its operands have been animated). This is all built on top of your `Syntax` package.

**Comment by mvdan** [comment]
> I much prefer your static parsing approach.

I still think doing the first change outlined above could be done. It does break full static parsing, but at least it wouldn't completely get rid of it. Do you have any thoughts? It would be useful to know if this run-time evaluation stuff happens often in the wild or not.

> I don't claim this is a good idea.

I agree this isn't a good idea :) It's basically the same as the option 2 explained earlier; parse the entire arithmetic expression as a word, expand its parts, split into fields, and only then evaluate the strings as an arithmetic expression.

> I'm writing my own to graphically animate bash in a very fine-grained way.

Oh, interesting. If you need that much control, then I agree that the `interp` package is not a good fit right now. It might be in the future.

You should still be able to use the `expand` package, though. That should save you a lot of code. That can do variable expansion, brace expansion, command substitution, arithmetic expansion, etc etc. If you need it to happen on a step-by-step or word-by-word basis, that should still be possible via its API.

**Comment by rob-myers** [comment]
Thanks, I'll take another look at `expand`.

I don't have strong feelings concerning `$(( $a ))`, but if it turns up I'll let you know.

The original question was concerned with `$(( $@ ))` which your parser can parse.
The problem was that `gosh` crashes -- do you know why?

**Comment by mvdan** [comment]
Ah yes - whatever we do, the interpreter shouldn't hang or crash. I'll look into that.

As for the behavior, I'll leave it as it is for now. We'll see if anyone else brings this up, or if this becomes a big deal in the future.
